plugins/pam.lua

```diff
@@ -0,0 +1,264 @@
+        return tonumber(context.configuration.gid_offset)
+    else
+        -- return default
+        return 100000
+    end
+end
+
+
+local get_config_uid_offset = function(context)
+    if context.configuration.uid_offset ~= nil then
+        return tonumber(context.configuration.uid_offset)
+    else
+        -- return default
+        return 100000
+    end
+end
+
+local get_config_default_group = function(context)
+    if context.configuration.default_group ~= nil then
+        return context.configuration.default_group
+    else
+        -- return default
+        return "pam_users"
+    end
+end
+
+local resolve_gid_default_group = function(context)
+    local group_name = get_config_default_group(context)
+    local groups = context.api:list_groups({ filter = "(cn=" .. group_name .. ")" })
+    local group_id = nil
+    if lldap.tables:empty(groups) then
+        lldap.log:debug("Creating group:" .. group_name)
+        local id, err = context.api:create_group({
+            display_name = group_name,
+            attributes = {}
+        })
+        if err ~= nil then
+            -- Error
+            error("Failed to create group: " .. group_name, 1)
+        end
+        lldap.log:debug("Created '" .. group_name .. "' group with id " .. tostring(id))
+        local group_gidnumber = get_config_gid_offset(context) + id
+        local res, err = context.api:update_group({
+            group_id = id,
+            insert_attributes = {
+                gidnumber = { int = group_gidnumber }
+            }
+        })
+        if err ~= nil then
+            error("Failed to set gidnumber attribute to '" .. group_name .. "' group", 1)
+        end
+        lldap.log:debug("Assigned gidnumber to '" .. group_name .. "': " .. tostring(group_gidnumber))
+        group_id = group_gidnumber
+    else
+        group_id = groups[1].attributes.gidnumber.int
+    end
+    return group_id
+end
+
+local resolve_next_uid = function(context, uid_offset)
+    local nextuid, err = context.kvstore:fetch_and_increment(KVKEYS.NEXT_UID, uid_offset)
+    if err ~= nil then
+        error("Unable to resolve next-uid from plugin KV store. Error: " .. tostring(err), 1)
+    end
+    return nextuid
+end
+
+local assign_user_attributes = function(context)
+    -- Ensure that the uidNumber for the created user is unique.
+    lldap.log:debug("Resolving current maximum uid and gid")
+    local uid_offset = get_config_uid_offset(context)
+    -- obtain list of all users in the system
+    local users, err = context.api:list_users({})
+    if err ~= nil then
+        -- Error
+        error("Failed to list_users: " .. tostring(err), 1)
+    end
+    if not lldap.tables:empty(users) then
+        -- obtain id of shared pam group
+        local group_id = resolve_gid_default_group(context)
+        lldap.log:debug("Resolved default group to ID: " .. tostring(group_id))
+        -- assign uid/gid to any users missing
+        for idx, user_and_group in pairs(users) do
```

Comment:
That _screams_ for a lock.

I mean, the worst case is that you'll skip over a number with 2 queries setting the "next uid" to the user, but still it's a great showcase of a lock.

Reply:
Don't disagree, but it is only for initialization. Once initialized they get the ID before creation with on_create_user, which only one node would proces.

But I agree, locks could be convenient. It seems like something that'd be suitable for a followup PR afterwards (I expect I'd still be up for doing it though).